[NeoHW] iP

NeoHW opened the pull request:
# Bytebuddy!!!

> "10/10 app - a loyal fan"

Bytebuddy frees your mind of having to remember things you need to do. It's,

* text-based
* Easy to learn
* ~FAST~ _SUPER_ FAST to use

All you need to do is, 
1. download it from [here](https://github.com/NeoHW/ip)
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :shushing_face:

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

How to use: 
1. Open Intellij (if you are not in the welcome screen, click `File` > `Close Project` to close the existing project first)
1. Open the project into Intellij as follows:
   1. Click `Open`.
   1. Select the project directory, and click `OK`.
   1. If there are any further prompts, accept the defaults.
1. Configure the project to use **JDK 11** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).<br>
   In the same dialog, set the **Project language level** field to the `SDK default` option.
1. After that, locate the `src/main/java/ByteBuddy.java` file, right-click it, and choose `Run ByteBuddy.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
```
   Hello from
   ____   __   __  _____  U _____ u   ____     _   _   ____     ____   __   __ 
U | __")u \ \ / / |_ " _| \| ___"|/U | __")uU |"|u| | |  _"\   |  _"\  \ \ / / 
 \|  _ \/  \ V /    | |    |  _|"   \|  _ \/ \| |\| |/| | | | /| | | |  \ V /  
  | |_) | U_|"|_u  /| |\   | |___    | |_) |  | |_| |U| |_| |\U| |_| |\U_|"|_u 
  |____/    |_|   u |_|U   |_____|   |____/  <<\___/  |____/ u |____/ u  |_|   
 _|| \\_.-,//|(_  _// \\_  <<   >>  _|| \\_ (__) )(    |||_     |||_ .-,//|(_  
(__) (__)\_) (__)(__) (__)(__) (__)(__) (__)    (__)  (__)_)   (__)_) \_) (__) 
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

NeoHW committed:
Level 0. Rename, Greet, Exit
(a) Give your chatbot a new name, to differentiate it from the placeholder name Duke. (ByteBuddy)
(b) Implement an initial skeletal version of the it that simply greets the user and exits.

NeoHW committed:
Level 1. Echo
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

NeoHW committed:
Level 2. Add, List
Add the ability to store whatever text entered by the user and display them back to the user when requested.

NeoHW committed:
Level 3. Mark as Done
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

1. Changed if-else to switch case
2. markAsDone() & unmarkAsDone() returns a string that can be piped into printWithSolidLineBreak()

NeoHW committed:
Level 4. ToDos, Events, Deadlines
Add support for tracking three types of tasks:

Level 4. ToDos, Events, Deadlines
Add support for tracking three types of tasks:
	ToDos: tasks without any date/time attached to it e.g., visit new theme park
	Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
	Events: tasks that start at a specific date/time and ends at a specific date/time e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

NeoHW committed:
abstracted method splitStringWithTrim that takes in the string information, and a string separator for future ease of reuse

NeoHW committed:
Command to do testing : ./runtest.sh
to run this script in text-ui-test directory ( if error try using sudo, should not need to)
A-TextUiTesting: Automated Text UI Testing

NeoHW committed:
Added DukeException class,
handled marking and unmarking index errors,
empty todo event deadline errors
wrong command errors

future improvements:
make sure deadline and event formatting / number of tokens input are correct

NeoHW committed:
1. Caught NumberFormatException e (e.g. mark abc instead of mark 1) , where user inputs string instead of int
2. Changed task status icon : done is ✓ and not done is x

NeoHW committed:
Level 6. Delete
Add support for deleting tasks from the list.

NeoHW committed:
Duke.java : Abstracted commands logic out in main and into methods	Avoid cluttering main by abstracting logic out into function and method calls

NeoHW committed:
changed global class variables to private static final

NeoHW committed:
Abstracted out main functionality into runByteBuddy method to reduce
main clutter

NeoHW committed:
fix switch case indentation

NeoHW committed:
minor changes

NeoHW committed:
Error messages variables
handle all errors possible (i hope)

NeoHW committed:
cleanup: removed unused thrown exceptions

NeoHW committed:
cleanup: remove unused thrown exception from main

Zer0Legion commented:
Insightful code!

NeoHW committed:
minor fix

NeoHW committed:
Add method to create data dir and output.txt file
This would handle the case where data file or dir does not exist at the start

NeoHW committed:
add functionality to save tasks in data/output.txt file

NeoHW committed:
Add functionality: Load data in TaskList from info in output.txt

NeoHW committed:
Merge branch 'branch-Level-7'

NeoHW committed:
Change split delimiters to use "/by" & "/from|/to" instead of just "/"
This fixes the issue when string info contains  "/" such as "2/12/2019 1800"

NeoHW committed:
Add support for LocalDate and LocalDateTime for Deadline class
Example: 2/12/2019 1800 = 2nd of December 2019, 6pm
2019-10-15 = Oct 15 2019

NeoHW committed:
Merge branch 'branch-Level-8'

NeoHW committed:
Add Date and DateTime support for Event class

NeoHW committed:
Refactor and divide classes into packages
Add TaskList class, Storage class

TODO: Add Ui and Parser class

NeoHW committed:
Add Ui class

NeoHW committed:
Bugfix: Fix "by from to" in deadline & event class

Bug: non date or datetime would not be captured

NeoHW committed:
Minor refactoring changes

NeoHW committed:
Add Parser class, Add command interface and specific command classes
Rename Duke to Bytebuddy

NeoHW committed:
Merge branch 'A-MoreOOP'

NeoHW committed:
Update README.md

NeoHW committed:
Update README.md

NeoHW committed:
ByteBuddy!

NeoHW committed:
Merge branch 'add-gradle-support' of https://github.com/NeoHW/ip

NeoHW committed:
Change application MainClassName in build.gradle to Bytebuddy
Previous value of seedu.duke.Duke leads to Error: Could not find or load main class seedu.duke.Duke

NeoHW committed:
Minor formatting changes

NeoHW committed:
Package files into src/main/java/bytebuddy package

NeoHW committed:
Gradle support

NeoHW committed:
Add TaskTest and TodoTest that tests Task and Todo

NeoHW committed:
Add TaskListTest file

NeoHW committed:
Bugfix: Remove mockito as it seems to interfere with gradle build and run

NeoHW committed:
Bugfix: `./gradlew clean build` runs tests and write testcases into output.txt
Changes to `build.gradle` to separate test sources from main sources

NeoHW committed:
Minor changes

NeoHW committed:
Add support for packaging app as a JAR file

NeoHW committed:
Add comment on where to find output JAR file
Info: JAR output is under build/libs
archiveBaseName = `duke`: This sets the base name of the generated JAR file to "duke". The resulting JAR file will be named something like "duke.jar".
archiveFileName = `bytebuddy.jar`: With this configuration, the generated JAR file will be named "bytebuddy.jar" instead of the default name based on archiveBaseName and archiveClassifier.
This is useful when you want more control over the naming of your artifacts. If you don't specify `archiveFileName`, the Shadow Plugin will generate a name based on `archiveBaseName`, `archiveAppendix`, `archiveVersion`, and `archiveClassifier`.

NeoHW committed:
Add ByteBuddy.java JavaDoc

NeoHW committed:
Add JavaDoc for commands

NeoHW committed:
Add JavaDoc for ByteBuddyException
Refactor DukeException -> ByteBuddyException

NeoHW committed:
Add JavaDoc for Parser, Storage and Ui

NeoHW committed:
Add JavaDocs for tasks

NeoHW committed:
Coding standard fix:
Names representing methods must be verbs

NeoHW committed:
Implement FindCommand
This allows users to find a task by searching for a keyword

NeoHW committed:
Merge tag 'A-CodingStandard'

NeoHW committed:
Merge tag 'Level-9'

NeoHW committed:
Merge tag 'A-JavaDoc'

# Conflicts:
#	src/main/java/bytebuddy/tasks/Task.java
#	src/main/java/bytebuddy/tasks/TaskList.java

NeoHW committed:
Tasklist: Change from System.out.println() to using Ui write methods
Change solidLineBreak to "-".repeat(60)

NeoHW committed:
Add CheckStyle config files

NeoHW committed:
Fix coding standards for main files using checkstyleMain

NeoHW committed:
Fix coding standards for test files using checkstyleTest

NeoHW committed:
Merge branch 'A-CheckStyle'

NeoHW committed:
Add feature: Find command allows for multiple keywords, separated by ","

NeoHW committed:
Bugfix: <command> <space> leads to CommandInfo being "" instead of emptyList

This leads to bugs such as `todo ` adding a todo task that has no description
Add LIST_FORMAT message to let users know how to use list command

NeoHW committed:
Merge branch 'A-Varargs'

NeoHW committed:
Update README.md

NeoHW committed:
Build.gradle: Add gradle dependencies

Create resources/images file,
added bytebuddy and user image

NeoHW committed:
Merge branch 'master' of https://github.com/NeoHW/ip into Level-10

NeoHW committed:
Add javaFX GUI for ByteBuddy chatbot

* Implement a graphical user interface (GUI) for the chatbot using
javaFX technology.
* Enhance user interaction by providing a visual interface
for input and output.
* This commit focuses on creating a more user-friendly experience,
promoting engagement.
* The use of JavaFX ensures a modern and responsive design
for the chatbot interface.

NeoHW committed:
Merge branch 'Level-10'

NeoHW committed:
Refactor: Enhance code readability and adhere to Checkstyle standards

This commit refactors the code in accordance with Checkstyle standards
to improve overall code readability and maintainability.

Key changes include:
- Proper indentation and spacing adjustments for consistency
- Ensured adherence to coding conventions outlined by Checkstyle rules
- Correct import classification and ordering

These changes contribute to a cleaner and more consistent codebase,
promoting better collaboration and long-term maintainability.

KinTatHo commented on src/main/java/bytebuddy/constants/ExceptionErrorMessages.java:
```diff
@@ -0,0 +1,12 @@
+package bytebuddy.constants;
```
this is a smart way to store different error messages

KinTatHo reviewed (changes requested):
Overall, really enjoyed reading your code. Code was understandable, learnt a couple of things from the code as well! Amazing job!

KinTatHo commented on src/main/java/bytebuddy/parser/Parser.java:
```diff
@@ -0,0 +1,77 @@
+        String commandType = commandParts[0].toLowerCase();
+        String commandInfo = commandParts.length > 1 ? commandParts[1].trim() : "";
+
+        switch (commandType) {
```
Nice job remembering the switch case indentation!

KinTatHo commented on src/main/java/bytebuddy/ByteBuddy.java:
Nice format, understandable

KinTatHo commented on src/main/java/bytebuddy/constants/Information.java:
```diff
@@ -0,0 +1,17 @@
+ * The {@code Information} class contains constants representing various informational strings used in the application.
+ */
+public class Information {
+    public static final String SOLID_LINE_BREAK = "-".repeat(90);
```
nice job remembering SCREAMING_SNAKE_CASE for constants

NeoHW committed:
Add assert statements to document important assumptions

The current situation lacks documentation of important assumptions in the code.
It is crucial to explicitly state these assumptions to improve code readability and maintainability.

Assert statements have been added at various points in the code to
document important assumptions that should hold true during runtime.

These assertions help clarify the expected behaviour and provide
immediate feedback if any assumptions are violated.

This approach is adopted to enhance code clarity and ensure that
future modifications or updates adhere to the documented assumptions.

Additionally, it aids in debugging by pinpointing potential issues early in the development process.

NeoHW committed:
Code quality refactoring

Some code contains deep nesting and lacks clear logical structure,
making it difficult to follow the flow of logic.

Deeply nested code and lack of logical structure decrease readability
and maintainability of the codebase.

The code is being restructured to improve readability.
Deep nesting is being avoided, and related segments of code are being
grouped logically using methods, indentation, and line spacing.

This follows best practices for code quality and readability.

NeoHW committed:
Merge pull request #1 from NeoHW/A-Assertions

Add assert statements to document important assumptions

NeoHW committed:
Merge branch 'master' into A-CodeQuality

NeoHW committed:
Merge pull request #2 from NeoHW/A-CodeQuality

Code quality refactoring

NeoHW committed:
Add GitHub Actions CI pipeline

The pipeline is triggered on both push and pull_request events.

The job performs the following steps:

* Sets up the repository by checking out the source code from the default branch.
* Explicitly checks out the source code from the master branch,
	ensuring consistency regardless of the default branch setting.
* Merges to the master branch by checking out the specific commit
	associated with the event.
* Validates the Gradle Wrapper.
* Sets up JDK 11 using the actions/setup-java action.
* Builds and checks the project using Gradle.

These changes aim to automate the build and testing process
for the project using GitHub Actions.

NeoHW committed:
Merge branch 'A-CI'

NeoHW committed:
Fix checkstyle issue and test CI workflow with GitHub Actions

Ensure that no line exceeds 120 characters per line to adhere to
commit message style guidelines and improve readability.

CI workflow failed due to a line exceeding 120 characters.
By resolving this issue, we ensure that the codebase complies with
the defined coding standards and that the CI workflow
successfully executes without errors.

Additionally, this commit demonstrates the effectiveness of the CI setup
by triggering a build and ensuring that the codebase is continuously
integrated and tested throughout the development process.

NeoHW committed:
Add duplicate task checking functionality

The current implementation lacks support for checking duplicate tasks,
leading to potential inconsistencies in task management.

To enhance the reliability and efficiency of task management,
it's essential to prevent the addition of duplicate tasks.
This ensures that tasks are unique and eliminates redundancy.

Additional changes:
Configure ./gradlew run to start GUI instead of text-based interface.

NeoHW committed:
Merge pull request #3 from NeoHW/C-DetectDuplicates

Add duplicate task checking functionality

NeoHW committed:
Update README.md

NeoHW committed:
Fix incorrect instruction format naming
Changed from LIST_FORMAT to FIND_FORMAT and
corrected instruction accordingly

NeoHW committed:
Add user guide for ByteBuddy chatbot

Created a user guide for new users to allow them to efficiently
use ByteBuddy chatbot to manage tasks.

Includes an image that shows how the GUI works, and a detailed
documentation of the various command and usage.
